{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b7223bb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import json\n",
    "OUTPUT_DIR = os.getenv(\"OUTPUT_DIR\")\n",
    "os.chdir(OUTPUT_DIR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7ff45223-03da-491c-8d66-7831c65088f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage Dir: /opt/brave_prod/workspace/analysis/9e8447b7-c496-4912-8da4-09a17ea0638f/eecd4ba4-4c1a-4876-aa91-ea233047bcd0/c9368753-9101-4211-ac83-c6feff7cf809\n"
     ]
    }
   ],
   "source": [
    "print(f\"Usage Dir: {OUTPUT_DIR}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cded0fa6-161d-4e72-8bff-aa6dfd4de88c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import shlex\n",
    "\n",
    "def run_command(cmd: str,shell=False):\n",
    "    \"\"\"\n",
    "    执行系统命令（支持实时输出），支持直接输入空格分割的字符串。\n",
    "    \"\"\"\n",
    "    # 自动按 shell 语法分割字符串，例如：\n",
    "    # \"ping -c 4 google.com\" -> [\"ping\", \"-c\", \"4\", \"google.com\"]\n",
    "    cmd_list = shlex.split(cmd)\n",
    "\n",
    "    process = subprocess.Popen(\n",
    "        cmd_list,\n",
    "        stdout=subprocess.PIPE,\n",
    "        stderr=subprocess.STDOUT,\n",
    "        shell=shell,   \n",
    "        text=True\n",
    "    )\n",
    "\n",
    "    for line in process.stdout:\n",
    "        print(line, end=\"\")  # 实时输出\n",
    "\n",
    "    process.wait()\n",
    "    print(f\"\\nExit code: {process.returncode}\")\n",
    "    return process.returncode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1bd604af-5b39-4a55-83ae-769651f19a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !ncbi-genome-download -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2f165753-a397-4bdf-a559-970c72019119",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"params.json\") as f:\n",
    "    params = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "f28fb62f-4364-468c-9cb3-39e65f56a6ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4791f5b8-96ec-4624-9edf-50ea3b03122b",
   "metadata": {},
   "outputs": [],
   "source": [
    "assembly_accessions = params[\"assembly_accessions\"]\n",
    "formats = params[\"formats\"]\n",
    "formats = \",\".join(formats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ab0b8359-c0a2-4032-9635-0927a10b3f1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'fasta,gff'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "b12c1a0b-9c1c-468d-8d56-5598f52b5f23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !ncbi-genome-download -v --taxids 511145  --assembly-levels complete --formats fasta bacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "8c619d8d-5300-4301-8cc4-f6992588d2ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Using cached summary.\n",
      "\n",
      "Checking assemblies:   0%|          | 0/1 [00:00<?, ?entries/s]INFO: Checking record 'GCF_000005845.2'\n",
      "\n",
      "Checking assemblies: 100%|██████████| 1/1 [00:00<00:00,  1.26entries/s]\n",
      "Checking assemblies: 100%|██████████| 1/1 [00:00<00:00,  1.26entries/s]\n",
      "\n",
      "Downloading assemblies: 0files [00:00, ?files/s]\n",
      "Downloading assemblies: 0files [00:00, ?files/s]\n",
      "\n",
      "Exit code: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd = f\"ncbi-genome-download -v  --assembly-accessions {assembly_accessions} bacteria --formats {formats} --progress-bar --flat-output  --output-folder output\"\n",
    "run_command(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "a022610e-3c49-46ab-8173-c67002ff3361",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: output/GCF_000005845.2_ASM584v2_genomic.fna already exists;\tnot overwritten\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "512"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system(f\"gunzip output/*.gz\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
